### PR TITLE
use cudnnRNNForwardInferenceEx for unpacked (padded) layout case

### PR DESCRIPTION
**Pull request by HectorSVC**
use cudnnRNNForwardInferenceEx for unpacked (padded) layout case so that we don't need extra kernel to process the seq_lens.

### CODE REVIEW

File: onnxruntime/core/providers/cuda/rnn/cudnn_rnn_base.cc [hunk 1]
```diff
@@ -288,12 +317,15 @@ Status CudnnRnnBase<T>::ComputeInternal(OpKernelContext* ctx) const {
     }
   }
 
-  if (sequence_lens_data != nullptr && y_h_data != nullptr && y_data != nullptr) {
+  if ((CUDNN_RNN_RELU == rnn_mode_ || CUDNN_RNN_TANH == rnn_mode_) && sequence_lens_data != nullptr && y_h_data != nullptr && y_data != nullptr) {
+    auto count = sequence_lens->Shape().Size();
+    auto sequence_lens_gpu = GetScratchBuffer<int32_t>(count);
+    cudaMemcpyAsync(sequence_lens_gpu.get(), sequence_lens_data, count * sizeof(int32_t), cudaMemcpyHostToDevice);
```
Comment (KeDengMS) [inline onnxruntime/core/providers/cuda/rnn/cudnn_rnn_base.cc]: >cudaMemcpyAsync [](start = 4, length = 15)

seems sequence_lens_data may be stomped by CPU before consumed by GPU? I think you need to create a temp CPU copy for it and do the deferred release.

Another question is why this is not caught be test...

File: onnxruntime/core/providers/cuda/rnn/cudnn_rnn_base.cc [hunk 2]
```diff
@@ -288,12 +317,20 @@ Status CudnnRnnBase<T>::ComputeInternal(OpKernelContext* ctx) const {
...
 
-  if (sequence_lens_data != nullptr && y_h_data != nullptr && y_data != nullptr) {
+  if ((CUDNN_RNN_RELU == rnn_mode_ || CUDNN_RNN_TANH == rnn_mode_) && sequence_lens_data != nullptr && y_h_data != nullptr && y_data != nullptr) {
+    auto count = sequence_lens->Shape().Size();
+    int device_id = 0;
+    CudaAsyncBuffer<int32_t> sequence_lens_buffer(this, device_id, count);
+    auto sequence_lens_cpu = sequence_lens_buffer.CpuSpan();
+    for (int i = 0; i < count; ++i) {
+      sequence_lens_cpu[i] = sequence_lens_data[i];
+    }
```
Comment (KeDengMS) [inline onnxruntime/core/providers/cuda/rnn/cudnn_rnn_base.cc]: Just memcpy?

File: onnxruntime/core/providers/cuda/rnn/cudnn_rnn_base.cc [hunk 3]
```diff
@@ -288,12 +317,20 @@ Status CudnnRnnBase<T>::ComputeInternal(OpKernelContext* ctx) const {
     }
   }
 
-  if (sequence_lens_data != nullptr && y_h_data != nullptr && y_data != nullptr) {
+  if ((CUDNN_RNN_RELU == rnn_mode_ || CUDNN_RNN_TANH == rnn_mode_) && sequence_lens_data != nullptr && y_h_data != nullptr && y_data != nullptr) {
+    auto count = sequence_lens->Shape().Size();
+    int device_id = 0;
```
Comment (KeDengMS) [inline onnxruntime/core/providers/cuda/rnn/cudnn_rnn_base.cc]: >device_id [](start = 8, length = 9)

You can get the device id from provider_. This won't work with multiple GPU

Comment (HectorSVC) [reply onnxruntime/core/providers/cuda/rnn/cudnn_rnn_base.cc]: I can fix it in this PR. and will create a work item to setup test environment and add test cases for multiple GPUs.

---
In reply to: [292754893](https://github.com/microsoft/onnxruntime/pull/899#discussion_r292754893) [](ancestors = 292754893)

Comment (KeDengMS) [reply onnxruntime/core/providers/cuda/rnn/cudnn_rnn_base.cc]: Thanks! LGTM

---
In reply to: [292757715](https://github.com/microsoft/onnxruntime/pull/899#discussion_r292757715) [](ancestors = 292757715,292754893)

**Pull request merged**